### ISSUE TITLE
Rotation not working

### ISSUE THREAD

**Issue by Adraesh** [SVGKit/SVGKit#148]
Hello,

We have an issue with an svg rotation. I use this method to convert the svg in UIImage :
UIImage\* newImage = [SVGKImage imageNamed:@"map-alaska-onlysimple.svg"].UIImage;

Here our code svg : http://pastebin.com/5U5XRmgT
the svg in image representation : http://adraesh.free.fr/svg_test.svg

Here what it's look like in UIImage :
http://adraesh.free.fr/uiimage.png

If someone has an idea ?

Thank you very much.

**Comment by adamgit** [comment]
Whats wrong with the image?

**Comment by Adraesh** [comment]
Hello adamgit,

The rotation is assumed to be at point 50, 50 which means in my case an image of 100 \* 100 in the middle and there as you can see it out of the frame. 

This is not normal

**Comment by Adraesh** [comment]
Here more explanations :

I've got an 100*100 SVG.

http://pastebin.com/TANy3H07

First thing, I have to put in my rotation -50 -50 I don't understand why but I tried everything and 50 50 did not work but the logic should be to assign 50 50 to a rotation at the center of an 100*100 image.

Here the display of my multiple SVG no rotation, it's perfect :
http://adraesh.free.fr/ios/no_rotate.png

Now if I apply a rotation and more I'am closer to 90 degrees, I lost my SVG. Rotate to 89°: 
http://adraesh.free.fr/ios/rotate_89.png

At 90 degrees :
http://adraesh.free.fr/ios/rotate_90.png

Above 90 degrees my SVG is displaying progressivelly
http://adraesh.free.fr/ios/rotate_91.png

To 270 degrees I lost my SVG again like in 90degrees:
http://adraesh.free.fr/ios/rotate_270.png

Above 270 degrees, my SVG is displaying progressivelly until a complete display at 360(no rotation)

Thank you again.

**Comment by adamgit** [comment]
The source code looks correct to me. Please run your rotation through the debugger, and try replacing the lines that do the rotate (if you search for "rotate" in the SVGkit project, there's only a couple of hits, it's easy to find). If that code is wrong (e.g. w.r.t. the -50,-50 you mention), try and work out why!

Re: the 90 / 270 thing, it looks like your line is too narrow to cover any pixels, once it has rotated. Try thickening the line - if that fixes it, that means it is a problem with "too narrow".

WHY it's too narrow: SVG's work in fractions-of-a-pixel, but UIImage only supports "whole" pixels. If you look at the code for generating a UIImage, we pass various options to Apple when telling them to make the UIImage - try changing them (I think some of them affect this).

QUICK FIX: if you shift your SVG by 0.5 pixels, that forces the line-boundaries to change, and someitmes makes "too narrow" pixels become "wide enough to draw".

**Comment by Adraesh** [comment]
When you talk about "look at the code for generating an UIImage" you talk about this line in the SVGKImage.m :

-(UIImage *)UIImage
{
    return [SVGKExporterUIImage exportAsUIImage:self antiAliased:TRUE curveFlatnessFactor:1.0f interpolationQuality:kCGInterpolationDefault]; // Apple defaults
}

**Comment by adamgit** [comment]
Yes - that method (exportAsUIImage:::) is doing the conversion work. The
arguments are things that we send to Apple to workaround other issues (to control exactly how Apple does the conversion).

If you change some of the arguments (e.g. interpolationQuality) they will
change "how" the bitmap is generated. You can experiment with them easily.

Better: cmd-click that method to jump to the source and see what it's
doing. There are additional options to Apple's code that we're currently
NOT using. Some of those PROBABLY affect the pixel-rendering in ways that
might fix your situation.

(NB: if we need extra arguments to "fix" your case, I'm happy to change
these method signatures and add the arguments. The current arguments were
chosen because "we needed them" - so it's OK to add more, if we also "need"
them :))

**Comment by Adraesh** [comment]
Ok adamgit, thanks alot.

Just a last question in this method :
http://pastebin.com/7yBzSz1Y

Where I have to cmd-click to see additionnals params that can be add ?

Thank you in advance.

**Comment by adamgit** [comment]
Look where the params are used. Keep cmd-clicking until you find a method
that is written by Apple.

Then read the Apple docs for that method / class / whatever. They will show
other methods / classes / options relating to that.

IIRC it's arguments to CGContextCreateBitmap... or similar (but read the
source to find out!)

**Comment by Adraesh** [comment]
Ok If I understand, everything happen in the renderToContext method.

So additionnal params can be add using the class CGContext :

https://developer.apple.com/library/mac/documentation/graphicsimaging/Reference/CGContext/Reference/reference.html

**Comment by Adraesh** [comment]
Ok adamgit ! I figured it out with my issue !

As I don't know what modify in your code I'm simply ask for a generation of a UIImage from a SVG to SVGKIT without rotation param. And after that I apply my custom rotation like this :

http://pastebin.com/HAjG5QQT

90 in the CGContextRotateCTM is my rotation in degrees.

So i can conclude that the issue comes from the rotation method.

**Comment by adamgit** [comment]
LOL yes, that would work too :).

But if there's a bug in SVGKit, would be nice to find and remove it.

**Comment by Adraesh** [comment]
Yes you right I will try to fix the rotation librairy code.

Thank you again.

**Comment by adamgit** [comment]
Did you make any progress with this? Did you find what was wrong with SVGKit?

**Comment by NasH5169** [comment]
Adam,

The rotation doesn't work with rotate(45,100,100) with a svg 200px / 200px, BUT ! :)

We can rotate with the matrix transform which works fine in SVGKit :

<g transform="matrix(0.7,0.7,-0.7,0.7,100,-40)">

cos(45), sin(45), -sin(45), cos(45), -100.cos(45) + 100.sin(45) + 100, -100.sin(45) - 100.cos(45) + 100

Adraesh and me will become some SVGKit masta !
